## Week 5 Homework

For this homework we will be using the FHV 2019-10 data found here. [FHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz)

In [1]:
# test script
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

print(spark \
      .range(1000 * 1000 * 1000) \
      .count())

1000000000


### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?

In [2]:
spark.version

'3.5.0'

### Question 2: 

**FHV October 2019**

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

- 1MB
- 6MB ✅
- 25MB
- 87MB

In [8]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz

--2024-03-01 14:30:47--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240301%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240301T143048Z&X-Amz-Expires=300&X-Amz-Signature=74d39d177f6a3f64d66e036aea09f2bc6d4d0601d4d377fa1f15f307cbaea76a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-03-01 14:30:48--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-

In [13]:
!gunzip fhv_tripdata_2019-10.csv.gz

In [22]:
! head --lines 10 fhv_tripdata_2019-10.csv > sample_fhv.csv

In [26]:
import pandas as pd

pd_df = pd.read_csv("sample_fhv.csv")
pd_df.dtypes

dispatching_base_num       object
pickup_datetime            object
dropOff_datetime           object
PUlocationID                int64
DOlocationID                int64
SR_Flag                   float64
Affiliated_base_number     object
dtype: object

In [30]:
from pyspark.sql import types

schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropOff_datetime', types.TimestampType(), True),
    types.StructField('PUlocationID', types.IntegerType(), True),
    types.StructField('DOlocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.FloatType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True)
])

In [31]:
df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv("fhv_tripdata_2019-10.csv")
df.head(5)

[Row(dispatching_base_num='B00009', pickup_datetime=datetime.datetime(2019, 10, 1, 0, 23), dropOff_datetime=datetime.datetime(2019, 10, 1, 0, 35), PUlocationID=264, DOlocationID=264, SR_Flag=None, Affiliated_base_number='B00009'),
 Row(dispatching_base_num='B00013', pickup_datetime=datetime.datetime(2019, 10, 1, 0, 11, 29), dropOff_datetime=datetime.datetime(2019, 10, 1, 0, 13, 22), PUlocationID=264, DOlocationID=264, SR_Flag=None, Affiliated_base_number='B00013'),
 Row(dispatching_base_num='B00014', pickup_datetime=datetime.datetime(2019, 10, 1, 0, 11, 43), dropOff_datetime=datetime.datetime(2019, 10, 1, 0, 37, 20), PUlocationID=264, DOlocationID=264, SR_Flag=None, Affiliated_base_number='B00014'),
 Row(dispatching_base_num='B00014', pickup_datetime=datetime.datetime(2019, 10, 1, 0, 56, 29), dropOff_datetime=datetime.datetime(2019, 10, 1, 0, 57, 47), PUlocationID=264, DOlocationID=264, SR_Flag=None, Affiliated_base_number='B00014'),
 Row(dispatching_base_num='B00014', pickup_datetime=

In [35]:
df = df.repartition(6)
df.write.parquet('fhv/')

In [40]:
!ls -lrth fhv

total 37M
-rw-r--r-- 1 jovyan users 6.1M Mar  1 15:45 part-00003-1fd7f3f2-d153-471c-90d7-5e4d8655f27b-c000.snappy.parquet
-rw-r--r-- 1 jovyan users 6.1M Mar  1 15:45 part-00001-1fd7f3f2-d153-471c-90d7-5e4d8655f27b-c000.snappy.parquet
-rw-r--r-- 1 jovyan users 6.1M Mar  1 15:45 part-00002-1fd7f3f2-d153-471c-90d7-5e4d8655f27b-c000.snappy.parquet
-rw-r--r-- 1 jovyan users 6.1M Mar  1 15:45 part-00004-1fd7f3f2-d153-471c-90d7-5e4d8655f27b-c000.snappy.parquet
-rw-r--r-- 1 jovyan users 6.0M Mar  1 15:45 part-00005-1fd7f3f2-d153-471c-90d7-5e4d8655f27b-c000.snappy.parquet
-rw-r--r-- 1 jovyan users 6.1M Mar  1 15:45 part-00000-1fd7f3f2-d153-471c-90d7-5e4d8655f27b-c000.snappy.parquet
-rw-r--r-- 1 jovyan users    0 Mar  1 15:45 _SUCCESS


### Question 3: 

**Count records** 

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

- 108,164
- 12,856
- 452,470
- 62,610 ✅

In [48]:
df  \
    .where("pickup_datetime >= '2019-10-15'") \
    .where("pickup_datetime < '2019-10-16'") \
    .count()

62610

### Question 4: 

**Longest trip for each day** 

What is the length of the longest trip in the dataset in hours?

- 631,152.50 Hours ✅
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

In [61]:
from pyspark.sql import functions as funcs

def hour_diff(time1, time2):
    timeFmt = "yyyy-MM-dd'T'HH:mm:ss.SSS"
    return (funcs.unix_timestamp(time1, format=timeFmt) - funcs.unix_timestamp(time2, format=timeFmt))/3600

df \
    .withColumn("trip_duration", hour_diff(df.dropOff_datetime, df.pickup_datetime)) \
    .agg(funcs.max('trip_duration')) \
    .show()

+------------------+
|max(trip_duration)|
+------------------+
|          631152.5|
+------------------+



### Question 5: 

**User Interface**

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- 4040 ✅
- 8080

### Question 6: 

**Least frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?</br>

- East Chelsea
- Jamaica Bay ✅
- Union Sq
- Crown Heights North

In [62]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2024-03-01 16:21:03--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240301%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240301T162104Z&X-Amz-Expires=300&X-Amz-Signature=5070199a3b6d8bd71fcd6f020b0e76b220c924cee4ff7e0b1369c5c635fc05fa&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2024-03-01 16:21:03--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [68]:
df_zones = spark.read \
    .option('header', 'true') \
    .csv('taxi_zone_lookup.csv')

In [73]:
df.registerTempTable("trips")
df_zones.createOrReplaceTempView("zones")

In [80]:
spark.sql("""
        SELECT Zone, Count(1)
        FROM trips
        JOIN zones on zones.LocationID = trips.PUlocationID
        GROUP BY Zone
        Order BY 2 ASC LIMIT 1
""").show()

+-----------+--------+
|       Zone|count(1)|
+-----------+--------+
|Jamaica Bay|       1|
+-----------+--------+

